### DISCUSSION TITLE
classes/class_inputeventmousebutton

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#250] [User-contributed notes]
# classes/class_inputeventmousebutton

Inherits: InputEventMouse< InputEventWithModifiers< InputEventFromWindow< InputEvent< Resource< RefCounted< Object Represents a mouse button being pressed or released. Description: Stores informati...

https://docs.godotengine.org/en/stable/classes/class_inputeventmousebutton.html

**Comment by serhpo** [comment]
I can not find event.is_double_click() in godot 4.3

**Reply by tcibils** [reply]
Shouldn't we have the position where the mouse was when it was clicked, so that we can use that information?

**Reply by Murksman** [reply]
the double_click property fills the role of is_double_click(). The event also already has the position property as it inherits from InputEventMouse.

**Comment by agorangetek** [comment]
```gdscript
func _input(event:InputEvent) -> void:
	if event is InputEventMouseButton:
		if event.button_index == MOUSE_BUTTON_LEFT and event.double_click:
			print("do something")
```

**Reply by serhpo** [reply]
I apologize, I don’t know why double click wasn’t found before, now everything is found, thank you all. 
event.double_click

**Reply by Marksstiker** [reply]
thank you so much

**Comment by tcibils** [comment]
Shouldn't we have the position where the mouse was when it was clicked, so that we can use that information?

**Reply by agorangetek** [reply]
It's in the event.

```gdscript
func _input(event:InputEvent) -> void:
	if event is InputEventMouseButton:
		if event.button_index == MOUSE_BUTTON_LEFT and event.double_click:
			print(event.position)

**Comment by Xionical1** [comment]
Just so everyone knows, code completion does not work for same line conditions.
```
if event is InputEventMouseButton and event.button_index == 1:
```
The editor does NOT show event.button_index to be a valid property.

**Reply by Murksman** [reply]
I agree. GDScript should eventually be changed to reflect clause class restriction in the conditional and anywhere in the inner block itself as the object is forced to inherit these properties. Im not sure how easy that would be however because of the potential edge cases.

**Reply by locust-of-chiron** [reply]
Isn't this intended behavior? The way this line should execute "under the hood" is evaluating each of the conditionals, &ing them together, and then jumping per the if switch.  If event isn't known to be an InputEventMouseButton, there is no guarantee that it even has a property of button_index.  To evaluate it left-to-right the way you're imagining, you have to use sequential ifs:
```
if event is InputEventMouseButton:
     if event.button_index == 1:
```

**Reply by Calinou** [reply]
[Lazy evaluation](https://en.wikipedia.org/wiki/Lazy_evaluation) makes it so `if event is InputEventMouseButton and event.button_index == 1` can be interpreted the same way as two nested `if` statements. The interpreter will not even evaluate the second condition if the first one evaluates to  `false`.

If the statement before the `and` evaluates to `false`, this means the statement after the `and` can have property access that would be otherwise invalid, or other "forbidden" operations that can't be detected at compile-time without actually causing runtime errors.

**Comment by hungptsg** [comment]
The code found on comments above mostly simulated a "mouse up" event. The full implementation that has "mouse up", "mouse down", "mouse clicked" should look like this.

```cs
using Godot;

public partial class DemoSimpleMouseOperations : Area2D {
  bool wasPressedPreviously = false;
  bool wasDoubleClickedPreviously = false;

  public override void _Ready() {
    InputEvent += (viewport, @event, shape_idx) => HandleInput(@event); ;
  }

  public void HandleInput(InputEvent ipev) {
    if (ipev is InputEventMouseButton mb) {
      // detect click
      if (mb.Pressed) {
        GD.Print("mouse down");
      }
      else if (!mb.Pressed) {
        GD.Print("mouse up");
        if (wasPressedPreviously) {
          GD.Print("mouse clicked");
        }
      }

      // detect double click
      if (wasDoubleClickedPreviously && !mb.DoubleClick) {
        GD.Print("mouse dbl-clicked");
      }
      wasPressedPreviously = mb.Pressed;
      wasDoubleClickedPreviously = mb.DoubleClick;
    }
  }
}
```
